source/s3_buffer_pool.c

```diff
@@ -0,0 +1,228 @@
+        buffer_pool->current_mem_usage -= buffer_pool->block_size;
+        aws_array_list_pop_back(&buffer_pool->blocks);
+    }
+}
+
+static struct aws_s3_pooled_buffer s_primary_acquire(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    uint8_t *alloc_ptr = NULL;
+    aws_mutex_lock(&buffer_pool->mutex);
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_count == 0) {
+            block->offset_ptr = block->block_ptr + size;
+            alloc_ptr = block->block_ptr;
+            block->alloc_count += 1;
+            break;
+        }
+
+        if (block->offset_ptr + size < block->block_ptr + block->block_size) {
+            alloc_ptr = block->offset_ptr;
+            block->offset_ptr += size;
+            block->alloc_count += 1;
+            break;
+        }
+    }
+
+    if (alloc_ptr == NULL && (buffer_pool->max_mem_usage == 0 ||
+                              buffer_pool->current_mem_usage + buffer_pool->block_size <= buffer_pool->max_mem_usage)) {
+        struct s3_buffer_pool_block block;
+        block.alloc_count = 1;
+        block.block_ptr = aws_mem_acquire(buffer_pool->base_allocator, buffer_pool->block_size);
+        block.offset_ptr = block.block_ptr + size;
+        block.block_size = buffer_pool->block_size;
+        aws_array_list_push_back(&buffer_pool->blocks, &block);
+        buffer_pool->current_mem_usage += buffer_pool->block_size;
+        alloc_ptr = block.block_ptr;
+    }
+
+    aws_mutex_unlock(&buffer_pool->mutex);
+
+    return (struct aws_s3_pooled_buffer){.ptr = alloc_ptr, .size = size};
+}
+
+struct aws_s3_pooled_buffer aws_s3_buffer_pool_acquire_buffer(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    AWS_PRECONDITION(buffer_pool);
+
+    if (size < buffer_pool->block_size) {
+        return s_primary_acquire(buffer_pool, size);
+    }
+
+    uint8_t *alloc_ptr = NULL;
+    aws_mutex_lock(&buffer_pool->mutex);
+    if (buffer_pool->max_mem_usage == 0 || buffer_pool->current_mem_usage + size <= buffer_pool->max_mem_usage) {
+        buffer_pool->current_mem_usage += size;
+        alloc_ptr = aws_mem_acquire(buffer_pool->base_allocator, size);
+    } else {
+        aws_s3_buffer_pool_trim(buffer_pool);
+        if (buffer_pool->current_mem_usage + size <= buffer_pool->max_mem_usage) {
+            buffer_pool->current_mem_usage += size;
+            alloc_ptr = aws_mem_acquire(buffer_pool->base_allocator, size);
+        }
+    }
+    aws_mutex_unlock(&buffer_pool->mutex);
+
+    return (struct aws_s3_pooled_buffer){.ptr = alloc_ptr, .size = alloc_ptr == NULL ? 0 : size};
+}
+
+void aws_s3_buffer_pool_release_buffer(
+    struct aws_s3_buffer_pool *buffer_pool,
+    struct aws_s3_pooled_buffer pooled_buffer) {
+    if (buffer_pool == NULL || pooled_buffer.ptr == NULL || pooled_buffer.size == 0) {
```

Comment:
I don't see us doing a size == 0 check in acquire, so if somehow someone passed size 0 we'd end up with mismatched refcounts

either eliminate this size == 0 check, or add one in acquire()